internal/util/bson.go

```diff
@@ -35,3 +38,37 @@ func (b *BSONArraySizer) Len() int {
 	// 4 bytes for BSON length, the elems, then a final NUL.
 	return 5 + b.elemsSize
 }
+
+// SplitArrayByBSONMaxSize takes an array of arbitrary Go values and
+// groups them so that, when built into a BSON array, each group exceeds
+// maxSize by the smallest length possible.
```

Comment:
(this feels like a weird definition of "max size", in that every element will exceed it!)

Reply:
Well, every element _but_, we assume, the last.

I wasn’t quite sure what to call it … `softMaxSize`?